Monika-After-Story/MonikaModDev
# Submod Framework

Introducing the submod framework
~~Finally making use of this project~~
(Big thanks to @Booplicate for helping me get this working)

# Adds:
- Submod class
- Submod settings screen

Finally, a system where submods can check for dependencies, have a general settings menu, and have their own update scripts.

## Testing:
first, make a new rpy, just put this in for the sake of example:
```renpy
screen submod_screen():
    vbox:
        box_wrap False
        xfill True
        xmaximum 1000


        hbox:
            style_prefix mas_ui.cbx_style_prefix
            box_wrap False

            textbutton _("Switch setting #1") action NullAction()
            textbutton _("Switch setting #2") action NullAction()
            textbutton _("Switch setting #3") action NullAction()
```
This allows us to generically add the same screen to the new submods settings screen.

- Make an rpy file and initialize a `Submod` object in it. Storing it doesn't matter as you shouldn't need to interact with it after.
```python
init -990 python in mas_submod_utils:
    Submod(
        "author",
        "generic name",
        "a description",
        "0.0.0"
    )
```
(this will do for that instruct)
- Verify that update scripts work for submods individually (pass in a dict similar to our update structs (follow the notes in `__init__` if you're confused. It's basically a {"from version": "to version"} system and it follows a chain up. Start with the version you initialized the submod object with initially)
```python
{
    "<author>_<name>_v<version>": "<author>_<name>_v<version>"
}
```
- Say goodbye, increment the version number, and load in again. The label stored in the value should now be seen.
- Assuming you haven't passed in `"submod_screen"` into `settings_pane`, verify that the `Submods` option does not show up in the settings screen (on the side with all the tabs)
- Say goodbye, pass `"submod_screen"` into the `settings_pane` param of the `Submod` and load in again. The `Submods` tab should be visible, selecting it should show that screen.
- Add a dependency to the submod (you can make up the name) pass in a value for the dependency in the following form: ("minimum_version", "maximum_version") (Maximum can be none)
- If you do not initialize another submod object with the required fields of the dependency (not installed/not recent enough version/too new version) verify that you get the correct tracebacks for each.

NOTE: This is likely confusing as hell, so ask me if you need help with this
